### PR TITLE
Bug 485625 - Allow overriding the conversion of a String into a Crede…

### PR DESCRIPTION
**Pull request by gouessej**
…ntial. First implementation

### CODE REVIEW

**Comment by eclipsewebmaster** [comment]
Issue tracker reference:
https://bugs.eclipse.org/bugs/show_bug.cgi?id=485625

File: jetty-util/src/main/java/org/eclipse/jetty/util/security/CredentialFactory.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+ * Class responsible for the creation of authentication credentials from strings.
+ * It supports both the known Credential types (CRYPT and MD5) and some 
+ * Credential types discovered at runtime by using the service loader mechanism.
+ *
+ * @see Credential
+ * @see Password
+ */
+public class CredentialFactory
+{
+	private ServiceLoader<CredentialProvider> credentialProviderLoader;
```
Comment (gregw) [inline jetty-util/src/main/java/org/eclipse/jetty/util/security/CredentialFactory.java]: spaces not tabs please

File: jetty-util/src/main/java/org/eclipse/jetty/util/security/CredentialFactory.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+	/**
+	 * Constructor
+	 */
+    public CredentialFactory() {
```
Comment (gregw) [inline jetty-util/src/main/java/org/eclipse/jetty/util/security/CredentialFactory.java]: curlies on new lines please. See the [coding standards](https://www.eclipse.org/jetty/documentation/current/coding-standards.html)

File: jetty-util/src/main/java/org/eclipse/jetty/util/security/CredentialFactory.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+     * @param credential String representation of the credential
+     * @return A Credential or Password instance.
+     */
+    public Credential getCredential(String credential) {
+    	if (credential.startsWith(Crypt.__TYPE)) return new Crypt(credential);
+        if (credential.startsWith(MD5.__TYPE)) return new MD5(credential);
+    	
+        for (CredentialProvider cp : credentialProviderLoader) {
+            if (credential.startsWith(cp.getPrefix())) {
+                return cp.getCredential(credential);
```
Comment (gregw) [inline jetty-util/src/main/java/org/eclipse/jetty/util/security/CredentialFactory.java]: perhaps allow for a null return here and then to keep trying other providers, so you can have multiple providers for the same prefix.

File: jetty-util/src/main/java/org/eclipse/jetty/util/security/CredentialFactory.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+     * known Credential type (eg "CRYPT:" or "MD5:" ) then a Credential of that
+     * type is returned. Otherwise, it tries to find a credential provider 
+     * whose prefix matches with the start of the credential String. Else the 
+     * credential is assumed to be a Password.
+     * 
+     * @param credential String representation of the credential
+     * @return A Credential or Password instance.
+     */
+    public Credential getCredential(String credential) {
+    	if (credential.startsWith(Crypt.__TYPE)) return new Crypt(credential);
```
Comment (gregw) [inline jetty-util/src/main/java/org/eclipse/jetty/util/security/CredentialFactory.java]: These could be move to providers.... or left here.... no strong preference.

File: jetty-util/src/main/java/org/eclipse/jetty/util/security/Credential.java [hunk 5]
```diff
@@ -70,10 +72,7 @@
      */
     public static Credential getCredential(String credential)
     {
-        if (credential.startsWith(Crypt.__TYPE)) return new Crypt(credential);
-        if (credential.startsWith(MD5.__TYPE)) return new MD5(credential);
-
-        return new Password(credential);
+        return CREDENTIAL_FACTORY.getCredential(credential);
```
Comment (gregw) [inline jetty-util/src/main/java/org/eclipse/jetty/util/security/Credential.java]: Why have a separate credential factory class?   Why not just put the behaviour here?

**Comment by gouessej** [comment]
Hi

@gregw I'll use this formatter to fix the spaces and the curlies: http://git.eclipse.org/c/jetty/org.eclipse.jetty.admin.git/tree/jetty-eclipse-java-format.xml

I'll use git commit -s to pass the intellectual property validation.

I'll keep trying other providers if null is returned.

I'll leave "new Crypt(credential)" and "new MD5(credential)" where they are except if someone else has a strong preference.

I can put the behaviour into Credential.java without creating a new class (CredentialFactory.java) but I'd like to know @janbartel 's position about your suggestion.

I'll make those changes in about 34 hours when I'm at home or earlier if possible.

**Comment by gouessej** [comment]
Please look at this pull request: https://github.com/eclipse/jetty.project/pull/72

**Pull request closed without merging**